### ISSUE TITLE
Segmentation Faults in pachd and Pipeline storage Containers

### ISSUE THREAD

**Issue by dgeorg42** [pachyderm/pachyderm#6543]
**What happened?**:

I have a single pipeline deployed for performance testing. The glob pattern is "/", so the entire input repo is treated as a single datum.  The code in the pipeline simply copies all of the files from the input repo to the output repo. My performance test puts 10,000 files at a time into the input repo using "pachctl put file", waits for the pipeline to complete, then puts another 10,000 files into the input repo. It does this until a total of 200,000 files exist in the repo, at which point it stops. Therefore, the pipeline will run a total of 20 times, each time with more files than the previous run.

I started a run of these tests last night, and checked the results this morning. I discovered that both pachd and my pipeline's storage container had triggered restarts due to segmentation faults, not just once, but multiple times:

```
{pach 2.0.0-beta.1 : aws} / {k8s 1.16.13-dispatcher : aws} / {git 2.24.3 : pachyderm/master}
~/Documents/Work/Repos/pachyderm -->> kubectl get pods
NAME                           READY   STATUS    RESTARTS   AGE
etcd-0                         1/1     Running   0          12h
pachd-6ffcddcfbc-7cm6v         1/1     Running   13         12h
pipeline-benchtests-v3-bd945   2/2     Running   11         11h
postgres-0                     1/1     Running   0          12h
```

From the end of the previous pachd container log:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1dd8c77]

goroutine 162 [running]:
github.com/pachyderm/pachyderm/v2/src/internal/metrics.(*Reporter).internalMetrics.func1(0xc0003ad680, 0x295b178, 0xc001972b00, 0xc000713dd8, 0xc001972cc0, 0x0)
	/Users/nitin/Code/Pachyderm/pachyderm/src/internal/metrics/metrics.go:264 +0x157
github.com/pachyderm/pachyderm/v2/src/internal/metrics.(*Reporter).internalMetrics(0xc0003ad680, 0xc000713dd8)
	/Users/nitin/Code/Pachyderm/pachyderm/src/internal/metrics/metrics.go:364 +0x148
github.com/pachyderm/pachyderm/v2/src/internal/metrics.(*Reporter).reportClusterMetrics(0xc0003ad680)
	/Users/nitin/Code/Pachyderm/pachyderm/src/internal/metrics/metrics.go:149 +0x8e
created by github.com/pachyderm/pachyderm/v2/src/internal/metrics.NewReporter
	/Users/nitin/Code/Pachyderm/pachyderm/src/internal/metrics/metrics.go:46 +0x15b
```

From the end of the previous pipeline storage container log:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1dd8c77]

goroutine 103 [running]:
github.com/pachyderm/pachyderm/v2/src/internal/metrics.(*Reporter).internalMetrics.func1(0xc000561c80, 0x295b178, 0xc007031140, 0xc000461dd8, 0xc007031300, 0x0)
	/Users/nitin/Code/Pachyderm/pachyderm/src/internal/metrics/metrics.go:264 +0x157
github.com/pachyderm/pachyderm/v2/src/internal/metrics.(*Reporter).internalMetrics(0xc000561c80, 0xc000461dd8)
	/Users/nitin/Code/Pachyderm/pachyderm/src/internal/metrics/metrics.go:364 +0x148
github.com/pachyderm/pachyderm/v2/src/internal/metrics.(*Reporter).reportClusterMetrics(0xc000561c80)
	/Users/nitin/Code/Pachyderm/pachyderm/src/internal/metrics/metrics.go:149 +0x8e
created by github.com/pachyderm/pachyderm/v2/src/internal/metrics.NewReporter
	/Users/nitin/Code/Pachyderm/pachyderm/src/internal/metrics/metrics.go:46 +0x15b
```

Now, I did discover a flaw in the way that my testing waits for the pipeline to complete, as a result of which it occasionally proceeds to upload the next set of files while the pipeline is still running. Obviously, that's an issue with my test, but a side effect of that is that the pipeline was sometimes running multiple jobs concurrently. In fact, that's the state I found things in this morning when I did a "list job":

```
ID                               PIPELINE   STARTED      DURATION           RESTART PROGRESS  DL       UL       STATE   
3ba46c8bed964514a244cf5cfbf0f260 benchtests 2 hours ago  -                  0       0 + 0 / 0 0B       0B       running 
138883d672594ad88589dfcbe8f37807 benchtests 2 hours ago  -                  2       0 + 0 / 0 0B       0B       running 
ae05806f782847aca7132db89eea674c benchtests 2 hours ago  2 minutes          0       1 + 0 / 1 995.6MiB 995.6MiB success 
7203a7ef9af34984b5e85e8c05ab12a8 benchtests 3 hours ago  31 minutes         0       1 + 0 / 1 928.7MiB 928.7MiB success 
34495b8303524d7abdfff6baafa417c8 benchtests 4 hours ago  2 minutes          0       1 + 0 / 1 862.3MiB 862.3MiB success 
8d861a01bfdc4da6a9e38c73718a9982 benchtests 11 hours ago About an hour      2       1 + 0 / 1 795.9MiB 795.9MiB success 
fd4e68a4da884a899e1051a5897644b6 benchtests 11 hours ago About a minute     0       1 + 0 / 1 729.5MiB 729.5MiB success 
d4d235cc130f4e8ab3e25333410bc987 benchtests 11 hours ago About a minute     0       1 + 0 / 1 663.1MiB 663.1MiB success 
74d034c621eb48f6b9e60b03fe4753eb benchtests 11 hours ago About a minute     0       1 + 0 / 1 596.7MiB 596.7MiB success 
759fc9ac2b6f4a6cb2646425c7bb7a43 benchtests 11 hours ago About a minute     0       1 + 0 / 1 530.3MiB 530.3MiB success 
506e8274b62d4a50a244d4ae2a520ce4 benchtests 11 hours ago 56 seconds         0       1 + 0 / 1 464MiB   464MiB   success 
d4fb7105044d4faabb9a3198a78527f0 benchtests 11 hours ago 49 seconds         0       1 + 0 / 1 397.9MiB 397.9MiB success 
255dd9bce5c6466e8891a34444779a85 benchtests 11 hours ago 42 seconds         0       1 + 0 / 1 331.6MiB 331.6MiB success 
4a0a46c539c94768b33133287be222ea benchtests 11 hours ago 33 seconds         0       1 + 0 / 1 265.2MiB 265.2MiB success 
d666cf61a82248b48483693b3517d1d5 benchtests 11 hours ago 25 seconds         0       1 + 0 / 1 199MiB   199MiB   success 
460bc41366f54af49fbfa55239d050e9 benchtests 11 hours ago 16 seconds         0       1 + 0 / 1 132.7MiB 132.7MiB success 
8843010dad01438496b1d8e41257d126 benchtests 11 hours ago 1 second           0       0 + 1 / 1 0B       0B       success 
a8fce725128841baa68e8aa14000b208 benchtests 11 hours ago 10 seconds         0       1 + 0 / 1 66.39MiB 66.39MiB success 
382f882b8db8421eaf973498849983a5 benchtests 11 hours ago 10 seconds         0       1 + 0 / 1 53.07MiB 53.07MiB success 
cedccad7e69649008a09cadce249ee25 benchtests 11 hours ago 8 seconds          0       1 + 0 / 1 39.77MiB 39.77MiB success 
4ad45f493b5d4eff8d00e6d785cbe200 benchtests 11 hours ago 6 seconds          0       1 + 0 / 1 26.58MiB 26.58MiB success 
33839a874e134bcab204a2626dea07ce benchtests 11 hours ago 4 seconds          0       1 + 0 / 1 13.3MiB  13.3MiB  success 
```

As you'll have noticed from my "get pods" listing earlier, I only have one pipeline pod running, so it's not clear to me how multiple jobs of the same pipeline could even be running in parallel at all, but I bring this up in case it might be a contributing factor.

The other thing which may (or may not) be contributing: I didn't do anything special with respect to resource requests, so by default my pipeline containers only requested 64 MiB. However, they also didn't specify any limit, I didn't see any OOM errors, and the pachd container requested 2 GiB and still produced segmentation faults, so perhaps this isn't a contributing factor after all.

**Environment?**:
- Kubernetes version (use `kubectl version`): v1.20.4-eks-6b7464
- Pachyderm CLI and pachd server version (use `pachctl version`): 2.0.0-beta.1
- Cloud provider (e.g. aws, azure, gke) or local deployment (e.g. minikube vs dockerized k8s): AWS
- OS (e.g. from /etc/os-release):
- Others:

**Comment by brycemcanally** [comment]
The core issue appears to be a pretty simple issue with the metrics collection which is doing a list pipeline without collecting the details and attempting to access the details in the resulting pipeline infos. As far as the parallel job stuff is concerned, I think it is more likely that we didn't terminate the prior job properly (I see a place where that can happen if the pachd sidecar goes down). Some other weird things are the two jobs that took much longer than the others and the one job that looks like it skipped the datum. We should probably look into this run / workload more to make sure that there is not a separate issue there.

**Comment by jrockway** [comment]
Duplicate of https://github.com/pachyderm/pachyderm/issues/6485
